### PR TITLE
feat(interceptors): added support for interceptors

### PR DESCRIPTION
**Pull request by Mordred**
Added support for interceptors. Interceptor can intercept four states:
 - `request` - when the message is constructed, but before it is sent
 - `requestError` - when there is an error during message building
 - `response` - when the response has returned
 - `responseError` - when the response wasn't successful

fixes: #46

NOTE: I had to temporary disable 2 tests (see https://github.com/jasmine/jasmine-ajax/issues/111)

Take a look and tell me what you think.

### CODE REVIEW

**Comment by EisenbergEffect** [comment]
@bryanrsmith Can you take a look at this?

File: test/http-client.spec.js [hunk 1]
```diff
@@ -496,37 +516,54 @@ describe('http client', () => {
...
 
-      var promise = client.send(new HttpRequestMessage('GET', 'some/cool/url'));
+      var promise = client.send(new HttpRequestMessage('GET', 'some/cool/path'));
       promise.catch((response, reason) => {
         expect(response instanceof HttpResponseMessage).toBe(true);
         expect(response.responseType).toBe('abort');
         done();
       });
 
+      // This have to be called in the nextTick because when we reach this code, XHR isn't sent yet
```
Comment (bryanrsmith) [inline test/http-client.spec.js]: Can't we use a timeout?

File: src/http-client.js [hunk 2]
```diff
@@ -34,10 +34,27 @@ export class HttpClient {
...
     this.requestProcessorFactories.set(JSONPRequestMessage, createJSONPRequestMessageProcessor);
     this.pendingRequests = [];
+    this.interceptors = [];
     this.isRequesting = false;
   }
 
   /**
+   * Add new interceptor
+   *
+   * NOTE: Interceptors are stored in reverse order. Inner interceptors before outer interceptors.
```
Comment (bryanrsmith) [inline src/http-client.js]: This is great info, but it's an implementation detail that shouldn't appear in documentation. I'd move it into a normal code comment inside the method.

File: src/http-client.js [hunk 3]
```diff
@@ -34,10 +34,27 @@ export class HttpClient {
...
+   * This reversal is needed so that we can build up the interception chain around the
+   * server request.
+   *
+   * @method addInterceptor
+   * @param interceptor A interceptor class with 4 possible methods: "request", "requestError", "response", "responseError"
```
Comment (bryanrsmith) [inline src/http-client.js]: "An interceptor"

File: src/http-client.js [hunk 4]
```diff
@@ -93,13 +110,36 @@ export class HttpClient {
...
+      return processor.process(this, message).then(response => {
+        trackRequestEnd(this, processor);
+        return response;
+      }).catch(response => {
+        trackRequestEnd(this, processor);
+        throw response;
+      });
+    };
+
+    var chain = [ processRequest, undefined ];
```
Comment (bryanrsmith) [inline src/http-client.js]: I think this would be easier to understand if it were an array of key/value pairs instead of an array whose elements alternate handler types. If you switch it I think the logic below will become simpler, too. Eg, you can use `promise = promise.then(...chain.shift())`.

File: src/http-client.js [hunk 5]
```diff
@@ -93,13 +110,36 @@ export class HttpClient {
...
     }
 
-    promise = processor.process(this, message).then(response => {
-      trackRequestEnd(this, processor);
-      return response;
-    }).catch(response => {
-      trackRequestEnd(this, processor);
-      throw response;
-    });
+    processRequest = (message) => {
```
Comment (bryanrsmith) [inline src/http-client.js]: I think the problem we're going to run into here is that all transformers run inside the processor. Most of the client's API uses transformers to set request properties, so the message that the `request` and `requestError` handlers see will pretty much always be an empty message.

What do you think about breaking up the `processor.process` method so that the message is fully constructed by one method and then sent by another? That would let us apply the transformers & everything to build the complete message, then run the request interceptors, then send the message. I think it would make interceptors a lot more useful since they'd be able to inspect and modify the message with the data that will be sent. It would also let the `requestError` interceptors catch any errors caused by transformers.

File: src/http-client.js [hunk 6]
```diff
@@ -93,13 +110,36 @@ export class HttpClient {
...
+        return response;
+      }).catch(response => {
+        trackRequestEnd(this, processor);
+        throw response;
+      });
+    };
+
+    var chain = [ processRequest, undefined ];
+    // Apply interceptors
+    for (let interceptor of this.interceptors) {
```
Comment (bryanrsmith) [inline src/http-client.js]: I really like for..of, but I think we're trying to avoid using it because it makes babel generate a bunch of extra code to handle iterators, and we need to try to keep file size low wherever possible.

File: test/http-client.spec.js [hunk 7]
```diff
@@ -534,60 +571,188 @@ describe('http client', () => {
...
+        .configure(x => x.withBaseUrl(baseUrl))
+        .addInterceptor(interceptor);
+
+      client.get('some/cool/path')
+        .catch((response) => {
+          expect(response.isSuccess).toBe(false);
+          expect(interceptor.called).toBe(true);
+          done();
+        })
     });
```
Comment (bryanrsmith) [inline test/http-client.spec.js]: I think it'd be great to have a test that verifies execution order of interceptors when you have multiple registered.

File: src/http-client.js (hunk 2 above)
Comment (bryanrsmith) [inline src/http-client.js]: Actually, is order significant? It doesn't look like interceptors get a reference to inner interceptors, and they don't appear to have a mechanism to short-circuit inner interceptors.

**Comment by bryanrsmith** [comment]
Thanks for working on this! I was just wishing we had interceptors the other day. I think it's a great feature.

File: test/http-client.spec.js (hunk 1 above)
Comment (Mordred) [inline test/http-client.spec.js]: It doesn't work with timeout - maybe I'm missing something.

**Comment by Mordred** [comment]
@bryanrsmith I have modified the PR:

- moved XHR creation and transformation to `processor.createXHR(client, message)` - if you have better naming, let me know
- interceptors are now called after all transformations are applied
- fixed abort - message processor now track if user call the abort() - this should fix case when abort() is called during processing interceptors (before XHR actually created)

> Actually, is order significant?

Yes, I think interceptors order is important. E.g.:

```js
class AuthenticateInterceptor {
    request(message) {
        if (/* user is authenticated */) {
            // add some auth token
            return message;
        } else {
            throw new Error('not-authenticated');
        }
    }
}

class AuthorizeInterceptor {
    request(message) {
        // Here we are sure that user is authenticated, so we can test permissions
        if (/* user is authorized */) {
            return message;
        } else {
            throw new Error('not-authorized');
        }
    }
}

client.add(new AuthenticateInterceptor());
client.add(new AuthorizeInterceptor());
```

`AuthorizeInterceptor` will never be called if the user is not authenticated.

File: src/request-message-processor.js [hunk 8]
```diff
@@ -18,28 +18,39 @@ export class RequestMessageProcessor {
...
     this.XHRType = xhrType;
     this.transformers = transformers;
+    this.isAborted = false;
   }
 
   abort(){
-    //The logic here is if the xhr object is not set then there is nothing to abort so the intent was carried out
-    if(this.xhr){
+    // The logic here is if the xhr object is not set then there is nothing to abort so the intent was carried out
+    // Also test if the XHR has
```
Comment (bryanrsmith) [inline src/request-message-processor.js]: This comment seems incomplete

File: src/request-message-processor.js [hunk 9]
```diff
@@ -18,28 +18,39 @@ export class RequestMessageProcessor {
...
+    if(this.xhr && this.xhr.readyState !== XMLHttpRequest.UNSENT){
```
Comment (bryanrsmith) [inline src/request-message-processor.js]: What happens if you abort the request before it's sent?

Comment (bryanrsmith) [inline src/request-message-processor.js]: Oh, never mind. I saw your comment below :-)

**Comment by bryanrsmith** [comment]
Thanks @Mordred. This looks great to me. My only reservation with merging it is that we can't test error cases any more. I wonder if we can find a jasmine expert to make a suggestion.

**Comment by EisenbergEffect** [comment]
@zewa666 Any thoughts on how we might test this new feature?
@bryanrsmith Do you feel this API melds well with the existing configuration API? Or do we need to make changes there?

**Comment by bryanrsmith** [comment]
Good point. For consistency we probably should support adding interceptors via the `.configure()`. It should be a simple change to add a new method to the request builder, and then copy the interceptors off the builder the same way we copy transformers.

**Comment by EisenbergEffect** [comment]
@Mordred Would you be interested in adding the above mentioned configuration work to your PR? @bryanrsmith or I could guide you if you aren't quite sure what to do.

**Comment by Mordred** [comment]
How should be interceptors ordered when copying from builder?

**Comment by EisenbergEffect** [comment]
@Mordred Probably we would have something like this:

```
RequestBuilder.addHelper('withInterceptor', function(interceptor){
  return function(client, processor, message){
    var interceptors = message.interceptors || (message.interceptors = []);
    interceptors.unshift(interceptor);
  };
});
```

Does that make sense?

**Comment by zewa666** [comment]
@EisenbergEffect nope no idea, guess the solution is what was described by @Mordred in the Jasmine-Ajax repo, now somebody just needs to implement that :)

**Comment by Mordred** [comment]
I have moved interceptors from the client to the message as `message.interceptors` and added helper `withInterceptor()` as @EisenbergEffect proposed.

Also added PR: jasmine/jasmine-ajax#112

**Comment by EisenbergEffect** [comment]
Cool. I like how this is turning out. @bryanrsmith Can you do one more review of this with these changes?

**Comment by bryanrsmith** [comment]
Looks great to me!

**Comment by EisenbergEffect** [comment]
Ok, I'm seeing three failed tests here.

* http client send can parse request headers
* http client send can parse multiple request headers
* http client send can parse header values containing :

@bryanrsmith @Mordred Do you guys know anything about this. Was this introduced as part of this PR?

**Comment by EisenbergEffect** [comment]
All tests seem to pass before applying this PR. Any ideas on this?

**Comment by Mordred** [comment]
Can you be more specific? I clone this branch to new directory and run `npm install && jspm install`, then `gulp test` and all tests (except 2 skipped) are passing:

```
[20:50:09] Using gulpfile /private/tmp/http-client/gulpfile.js
[20:50:09] Starting 'test'...
INFO [karma]: Karma v0.12.35 server started at http://localhost:9876/
INFO [launcher]: Starting browser Chrome
WARN [web-server]: 404: /favicon.ico
INFO [Chrome 43.0.2357 (Mac OS X 10.10.3)]: Connected on socket FL2K6KySD8aHtSqynAJt with id 58396447
Chrome 43.0.2357 (Mac OS X 10.10.3): Executed 83 of 85 (skipped 2) SUCCESS (0.16 secs / 0.139 secs)
[20:50:14] Finished 'test' after 5.2 s
```

Can you write your output?

PS: It would be great to have something as Travis CI

**Comment by EisenbergEffect** [comment]
Interesting. I'm wondering if I need to update jasmine or something. Let me try again.
PS. We are working on getting CI set up :)

**Comment by EisenbergEffect** [comment]
It's a jasmine version issue. I got it sorted now. Merge coming shortly.

**Comment by Mordred** [comment]
I thought you will wait till all tests will be back (jasmine/jasmine-ajax#111 will be fixed)

**Comment by EisenbergEffect** [comment]
Is there an ETA on that?

**Comment by Mordred** [comment]
I don't know .. there is no response from jasmine team yet.

**Comment by ottomatic** [comment]
Hi,

We are looking at using Aurelia in a setting with OpenID + OAuth2 authentication and authorization. 

I'm guessing this interceptor feature is an important building block in order to smoothly authorize subsequent requests to a web api after authentication. Is this a correct assumption?
